Mathlib/Algebra/Polynomial/Sequence.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2025 Julian Berman. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Hill, Julian Berman, Austin Letson, Matej Penciak
+-/
+import Mathlib.Algebra.Polynomial.Monic
+import Mathlib.LinearAlgebra.Basis.Basic
+
+/-!
+
+# Polynomial sequences
+
+We define polynomial sequences – sequences of polynomials `a₀, a₁, ...` such that the polynomial
+`aᵢ` has degree `i`.
+
+## Main definitions
+
+* `Polynomial.Sequence R`: the type of polynomial sequences with coefficients in `R`
+
+## Main statements
+
+* `Polynomial.Sequence.basis`: a sequence is a basis for `R[X]`
+
+## TODO
+
+Generalize linear independence to:
+  * just require coefficients are regular
+  * arbitrary sets of polynomials which are pairwise different degree.
+-/
+
+open Submodule
+open scoped Function
+
+universe u
+
+variable (R : Type u)
+
+namespace Polynomial
+
+/-- A sequence of polynomials such that the polynomial at index `i` has degree `i`. -/
+structure Sequence [Semiring R] where
+  /-- The `i`'th element in the sequence. Use `S i` instead, defined via `CoeFun`. -/
+  protected elems' : ℕ → R[X]
+  /-- The `i`'th element in the sequence has degree `i`. Use `S.degree_eq` instead, -/
+  protected degree_eq' (i : ℕ) : (elems' i).degree = i
+
+attribute [coe] Sequence.elems'
+
+namespace Sequence
+
+variable {R}
+
+/-- Make `S i` mean `S.elems i`. -/
+instance coeFun [Semiring R] : CoeFun (Sequence R) (fun _ ↦  ℕ → R[X]) := ⟨Sequence.elems'⟩
+
+section Semiring
+
+variable [Semiring R] (S : Sequence R)
+
+/-- `S i` has degree `i`. -/
+@[simp]
+lemma degree_eq (i : ℕ) : (S i).degree = i := S.degree_eq' i
+
+/-- `S i` has `natDegree` `i`. -/
+@[simp]
+lemma natDegree_eq (i : ℕ) : (S i).natDegree = i := natDegree_eq_of_degree_eq_some <| S.degree_eq i
+
+/-- No polynomial in the sequence is zero. -/
+lemma ne_zero (i : ℕ) : S i ≠ 0 := degree_ne_bot.mp (by simp [S.degree_eq i])
+
+/-- No two elements in the sequence have the same degree. -/
+lemma degree_ne_degree {i j : ℕ} (h : i ≠ j) : (S i).degree ≠ (S j).degree := by
+  simp [S.degree_eq i, S.degree_eq j, h]
```

Comment:
Do we want a variant of this lemma with `Function.Injective`, maybe?

Reply:
Added (hopefully correctly).